lib/DXIL/DxilShaderModel.cpp

```diff
@@ -494,6 +494,20 @@ const llvm::StringRef ShaderModel::FullNameFromKind(DXIL::ShaderKind sk) {
   }
 }
 
+bool ShaderModel::AllowDerivatives(DXIL::ShaderKind sk) const {
+  switch (sk) {
+  case DXIL::ShaderKind::Pixel:
+  case DXIL::ShaderKind::Library:
+  case DXIL::ShaderKind::Node:
+    return true;
+  case DXIL::ShaderKind::Compute:
+  case DXIL::ShaderKind::Amplification:
+  case DXIL::ShaderKind::Mesh:
+    return IsSM66Plus();
```

Comment:
Now, since this checks shader model, but doesn't differentiate result between failure due to kind vs. model, the later code that would emit `warn_hlsl_derivatives_in_wrong_shader_model` is unreachable.  Perhaps now we should rename this function to AllowDerivativesInShaderKind, and there's no need for it to be a method on shader model.

Alternatively, we could encapsulate all the required logic in one diagnostic function (not as a method of ShaderModel), which can be called from the switch on derivative ops in `DiagnoseUsedHLSLMethodCall`.

If this logic is needed outside `SemaHLSLDiagnoseTU.cpp`, like for reuse in DxilValidation, then it could be given extracted values (such as for NodeLaunch and numthreads), located here, and could return an enum for the source of incompatibility with derivatives (translated to a diagnostic by the caller).

Reply:
Moved AllowDerivatives check into the switch.